### PR TITLE
Forex support through Oanda

### PR DESCRIPTION
**Pull request by alokan**
I used poetry export for requirements and requirements-full. I'm not sure why so many things changed.

### CODE REVIEW

File: gamestonk_terminal/oanda/oanda_functions.py
Comment (DidierRLopes) [inline gamestonk_terminal/oanda/oanda_functions.py:1]: This file should be called "fx_view.py" instead

**Review by DidierRLopes — COMMENTED** [review]
Commented on minor things. Good work!!

You also need to update main_helper where there's the main terminal print.
So that people know that there's a new `fx` menu.

Regarding the poetry, should be fine as you've done! 

I'll try to give this a go between today and tomorrow, will need to create a Oanda account. Also you may want to try to add a try-catch block that checks if user account is valid, because the terminal crashed when I accessed the menu without a valid account login

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:178]: here should be `cancel`

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:188]: Since that's the only parameter, you can add this:
```
if l_args:
       if "-" not in l_args[0]:
                l_args.insert(0, "-i")
```

So that the user can do: `cancel x`

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:194]: Can you add a `print("")` at end.

Also, when it fails, it should fail just with a simple error message to the user.

Captura de ecrã 2021-04-18, às 21 12 38

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:196]: Instead: `print(e, "\n")`

File: gamestonk_terminal/forex/fx_controller.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_controller.py:54]: here you don't need to follow alphabetical order.

You can re-structure this by categories so that is easier for the user to understand the organization. Up to you entirely :)

File: gamestonk_terminal/forex/fx_controller.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_controller.py:55]: I get this with `candles`:

```
(✨) (fx)> candles
{"errorMessage":"Invalid value specified for 'instrument'"}
```

and this is because I haven't loaded an instrument. So I think you should hide the commands that require an instrument. Similarly to how the main menu is working now, where it doesn't show Technical Analysis for instance because user hasn't load ticker.

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:361]: It should fail more gracefully than:
```
(✨) (fx)> candles
{"errorMessage":"Invalid value specified for 'instrument'"}
```

by printing the error message and a new line.

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:471]: print("") at the end.

I'll stop commenting this, but ideally all commands end up with a new line :)

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:299]: Here again, I would allow the user to do:
`close 3` by adding that code snipped that I mentioned before.

I'll stop commenting this forward, I'll leave it up to your judgment to add it. This is just to improve user experience.

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:308]: Instead of just printing response, you should check if there's an error, there, if there is just print the error string. This applies to all commands

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:280]: It looks weird, but in all commands `prog=` should have the name of the terminal option rather than the function, so that it appears correctly in `usage: `

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:72]: This one crashed badly for me:

```
  File "/Users/DidierRodriguesLopes/Documents/git/GamestonkTerminal/gamestonk_terminal/forex/fx_controller.py", line 83, in switch
    )(other_args)
  File "/Users/DidierRodriguesLopes/Documents/git/GamestonkTerminal/gamestonk_terminal/forex/fx_controller.py", line 118, in call_list
    fx_view.list_orders(account, other_args)
  File "/Users/DidierRodriguesLopes/Documents/git/GamestonkTerminal/gamestonk_terminal/forex/fx_view.py", line 96, in list_orders
    df = df[["id", "instrument", "units", "price", "state", "type"]]
  File "/Users/DidierRodriguesLopes/anaconda3/envs/gst/lib/python3.6/site-packages/pandas/core/frame.py", line 2912, in __getitem__
    indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
  File "/Users/DidierRodriguesLopes/anaconda3/envs/gst/lib/python3.6/site-packages/pandas/core/indexing.py", line 1254, in _get_listlike_indexer
    self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
  File "/Users/DidierRodriguesLopes/anaconda3/envs/gst/lib/python3.6/site-packages/pandas/core/indexing.py", line 1298, in _validate_read_indexer
    raise KeyError(f"None of [{key}] are in the [{axis_name}]")
KeyError: "None of [Index(['id', 'instrument', 'units', 'price', 'state', 'type'], dtype='object')] are in the [columns]"
```

File: gamestonk_terminal/forex/fx_controller.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_controller.py:53]: I would re-structure a bit this with the loaded instrument, doing something like:

```
print("    quit          Quit to abandon program")
print("")
print("Instrument: EUR/USD")
print("")
print("    load   ...")
```

File: gamestonk_terminal/forex/fx_view.py
Comment (DidierRLopes) [inline gamestonk_terminal/forex/fx_view.py:250]: This one also crashed badly for me. 

Even if there are no arguments, it would be good to still use argparse to add a definition to the command so that the user can do:
```
trades -h
```
and see "shows open trades" or something like that

**Review by DidierRLopes — CHANGES_REQUESTED** [review]
Overall, great work! Sorry for the late review, only now got around signing up for an account to test this!

I've got some minor points that I would like for this to be addressed before merging:

* Check for wrong api keys what happens. We have to try to ensure a clean experience to every user
* argparse for all commands so that the users can do `command -h`
* the `prog=` should have the name of the command rather than the name of the function
* When there is an error or so, don't print the dictionary, but print the important error message in string format to the user
* Always add a new line after a command, regardless of an error or not. To keep consistency across.
* In some commands it makes sense to allow user to do `load EUR/USD` instead of `load -i EUR/USD`. The snippet for this goes attached to my review
* Refactor fx_controller print help so that it makes more sense from a user perspective. E.g. adding the loaded instrument, and only printing some of the commands that rely on an instrument being loaded previously

Sorry for extensive review, but this will help users to have a smoother experience across the terminal! Great work!!

**Comment by alokan** [comment]
Okay I think I got everything. V20Error isn't subscriptable, so I had to slice the string to get the error message out. Some of the longer ones I didn't encounter may still need to be done. Let me know if you find any or if I missed anything.

**Comment by DidierRLopes** [comment]
> Okay I think I got everything. V20Error isn't subscriptable, so I had to slice the string to get the error message out. Some of the longer ones I didn't encounter may still need to be done. Let me know if you find any or if I missed anything.

Nice, thank you!!

Do you mind if I do a PR to your branch with the latest changes? Probably is easier that way

**Comment by alokan** [comment]
No, I don't mind.

**Comment by DidierRLopes** [comment]
Nice, thank you!

See https://github.com/alokan/GamestonkTerminal/pull/22

**Review by aia — APPROVED** [review]
Looks ready

**Comment by alokan** [comment]
Okay I added a README, and a function to change the seperator. It takes an argument to what you want it changed to because stocktwits bullbear doesn't return anything for EUR_USD but does for EUR/USD. Also most posts on reddit have a / as the currency separators. Those aren't in this PR but will be when I get time to test what other parts of the terminal can be integrated and which formats return the most information. I also changed the default count for list_orders to 20 so it doesn't overrun the screen with past orders, added a check for impact from a key error that was raised when calling the calendar to see past events.

File: README.md
Comment (DidierRLopes) [inline README.md:262]: I think you can leave this comment.

File: gamestonk_terminal/forex/README.md
Comment (DidierRLopes) [inline gamestonk_terminal/forex/README.md:41]: *calendar

The link is broken because there's a typo

File: gamestonk_terminal/forex/README.md
Comment (DidierRLopes) [inline gamestonk_terminal/forex/README.md:32]: indentation

File: gamestonk_terminal/config_terminal.py
Comment (DidierRLopes) [inline gamestonk_terminal/config_terminal.py:39]: Instead, can you write: https://developer.oanda.com

**Review by DidierRLopes — COMMENTED** [review]
Things are looking good! This is the last round of review.

Btw, would you be happy if someone asked for a feature regarding Forex for us to ping you so that you keep maintaining it?

**Comment by alokan** [comment]
> Btw, would you be happy if someone asked for a feature regarding Forex for us to ping you so that you keep maintaining it?

Yeah that would be great. I still plan on adding a few things. I just wanted to get your opinions on what was there so far and what needed to be changed.

**Comment by DidierRLopes** [comment]
Can you resolve the conflicts before merging?

**Review by DidierRLopes — APPROVED** [review]

**Comment by DidierRLopes** [comment]
Amazing work!! A lot of people will appreciate your effort with this. Thank you!

**Pull request merged**
